auth: add Casbin as an authorization module for v3 API.

hsluoyz opened the pull request:
This uses [Casbin](https://github.com/casbin/casbin) as another auth store for v3 API. It provides ACL, RBAC, ABAC authorization for v3 API by providing flexible policy definition.

Fixes https://github.com/coreos/etcd/issues/7979

heyitsanthony commented on auth/casbin_store.go:
```diff
@@ -0,0 +1,224 @@
+}
+
+func NewCasbinAuthStore(be backend.Backend, tp TokenProvider) *casbinAuthStore {
+	e := casbin.NewEnforcer("casbin_model.conf", "casbin_policy.csv")
```
Is the intent that `casbin_policy.csv` holds user-configurable policies? If so, how is this configuration synchronized across all members? If one member rejects but another accepts then the cluster will be inconsistent,

hsluoyz replied:
Yeah, the policy rules should be stored in etcd itself as the way the default auth store does. I will work on it.

hsluoyz replied:
Hi @heyitsanthony , the user-configurable policies are now store in etcd backend. Please see the update of this PR.

mitake commented:
@hsluoyz I looked around this PR and the basic strategy of changing seems to be ok. But I want to check that how casbin works in etcd. Is it possible for you to update this PR for demonstration? At least adding a new option `--auth-backend` and switching the auth stores during the initialization would be required.

hsluoyz commented:
Hi @mitake ,

I have added the option ``--auth-backend``. Please use ``--auth-backend casbin`` to enable the Casbin auth store.

To see how Casbin auth store works, you can also enable the Casbin auth store's Verbose mode by changing ``auth/casbin_store.go:300`` from:

```go
e := casbin.NewEnforcer(m, a, false)
```

to:

```go
e := casbin.NewEnforcer(m, a, true)
```

It will print more logs to the console.

mitake commented:
@hsluoyz I couldn't try casbin because some part of casbin store depends on the default auth store e.g. `AuthEnable()` of casbin store directly calls `AuthEnable()` of the default store, and the default store's `AuthEanble()` calls its `getUser()` (https://github.com/coreos/etcd/blob/master/auth/store.go#L805), the `getUser()` looks up a user from the bucket of the default store that differs from `enforcer` of casbin. It means that some part of the information that is required for the auth feature is store in casbin and other part is stored in the default store.

Could you fix the problem? Ideally `e2e/ctl_v3_auth_test.go` should be executed with the casbin store.

hsluoyz commented:
Hi @mitake ,

Yes, currently Casbin auth store borrows some code like ``AuthEnable()`` from the default auth store because there are some shared behaviors. But I don't quite understand: ``It means that some part of the information that is required for the auth feature is store in casbin and other part is stored in the default store.``

From the definition of ``NewCasbinAuthStore()``, we can see that the two parameters ``be backend.Backend, tp TokenProvider`` are shared between Casbin auth store and the default auth store wrapped inside Casbin auth store. So in fact the data are still finally stored inside the same backend ``be``.

The only difference is the default auth store uses three buckets: ``auth, authUsers, authRoles``. And Casbin auth store uses four buckets: the above three buckets and an additional ``casbinAuth`` bucket which is used to store Casbin rules. IMHO, these buckets are all stored in the same backend ``be``, right? Because the code is like:

```go
tx := be.BatchTx()
tx.Lock()

tx.UnsafeCreateBucket(casbinAuthBucketName)
```

So I don't quite understand what the store means in ``looks up a user from the bucket of the default store that differs from enforcer of casbin.``. Is it the ``casbinAuthStore`` Go struct, or the backend ``be``?

If you mean the Go struct, I know the current implementation is not the ultimate form. Because "casbin" and "default" should be parallel options for auth mode. And one shouldn't depend on another. But there's indeed some code that Casbin auth store shares with default auth store. I don't think making a copy for that code is good. My thought is to refactor the shared code out as an interface that all auth stores (default, or casbin, or another new store) should follow. And we should even refactor out an authentication store to handle things like user-password checking. If this is what you mean, do you think I should do the refactoring in this PR together or in following efforts?

If you mean the backend ``be``, then I think Casbin auth store and the default auth store already share the same backend, as it is passed as a parameter in the New method.

Thanks.

mitake commented:
@hsluoyz casbin stores its user information (e.g. which roles the user belong to) in its internal state, right? And the state can only be accessed by `enforcer.GetRolesForUser()` or `enforcer.AddRoleForUser()`. So if etcd is using casbin as a backend, it shouldn't interpret the content of `authpb.User.Roles`. Current implementation of casbin backend depends on the field because it calls `getUser()` of `auth/store.go` e.g. in a case of `AuthEnable()`. I modified slightly like this: https://github.com/mitake/etcd/commit/263dfa34f76b6f4a411fcd04bb718bbc80ad54e0#diff-dea7aee7def796e871ca415b9e8bbd28R805 and without the change, I couldn't enable auth with casbin. BTW the change is just for demonstration so please don't follow the manner ;)

(The reason why `getUser()` matters for the enabling is that etcd assumes that there is a `root` user that has `root` role before enabling)

I suggest to test enable casbin based auth of working etcd cluster. Possibly this script is useful for you for preparation: https://gist.github.com/mitake/c86ddc353929d1663c7b102282f7dc8e . After executing the script, you can enable auth with `ETCDCTL_API=3 etcdctl auth enable`.

hsluoyz commented:
@mitake , 

> casbin stores its user information (e.g. which roles the user belong to) in its internal state, right? And the state can only be accessed by enforcer.GetRolesForUser() or enforcer.AddRoleForUser().

This is correct:) 

> So if etcd is using casbin as a backend, it shouldn't interpret the content of authpb.User.Roles. Current implementation of casbin backend depends on the field because it calls getUser() of auth/store.go e.g. in a case of AuthEnable().

Casbin auth store calls ``getUser()`` only to check if the user exists, it doesn't read or write the value of ``authpb.User.Roles``. So Casbin auth store doesn't depend on the data of default auth store. In fact, if Casbin auth store is enabled, ``authpb.User.Roles`` will always be empty because it's not assigned.

For example, I searched the usage of ``authpb.User.Roles``. They are:

- getMergedPerms --> not used by Casbin auth store
- hasRootRole --> overrided by Casbin auth store
- RoleDelete --> overrided by Casbin auth store (**I found a bug here, Casbin should delete related user-role mapping when role is deleted**, have fixed)
- UserChangePassword --> not used by Casbin auth store
- UserGet --> overrided by Casbin auth store, and already replaced the ``.Roles`` with Casbin roles.
- UserGrantRole --> overrided by Casbin auth store
- UserRevokeRole --> overrided by Casbin auth store

So Casbin auth store has no interaction with ``authpb.User.Roles``.

In my thought, Casbin auth store calling ``getUser()`` is only a workaround for now. Because checking whether a user exists belongs to authentication and should be refactored out from the current authorization store. If you agree, I can do the refactoring work in next PRs.

> without the change, I couldn't enable auth with casbin. BTW the change is just for demonstration so please don't follow the manner ;)

There are some bugs in original code. I believe I have fixed them in the updates. Please try again!

> (The reason why getUser() matters for the enabling is that etcd assumes that there is a root user that has root role before enabling)

I have removed the root role dependence in Casbin auth store's ``AuthEnable()`` method in latest update. Now it only requires to create a root user. So the current behavior is:

- root user can do anything.
- If root role is created and granted to a user, that user can do anything too. But the root user doesn't rely on the root role any more.

I think it has simplified the setup because we don't need to enforce to create a root role, and grant root role to root user.

> I suggest to test enable casbin based auth of working etcd cluster. Possibly this script is useful for you for preparation: https://gist.github.com/mitake/c86ddc353929d1663c7b102282f7dc8e . After executing the script, you can enable auth with ETCDCTL_API=3 etcdctl auth enable.

I tried this script. It works fine. Thanks!

heyitsanthony commented:
I'd really prefer to have this implemented as a proxy first instead of making casbin a hard dependency for the etcd server.

hsluoyz commented:
Hi @heyitsanthony , I think we should go step by step. What you said about LDAP is really authentication instead of authorization. Authentication cares about who you are. Authorization cares about what you can do. They are different things. The current design mixed authentication and authorization and they should be separated sooner or later. I can do it in following PRs.

This PR only follows the design of the default auth store, and tries to provide the users another option similar to the default auth. What you said about proxy may be a long-term goal, but I didn't see any efforts towards it though. Or can you port the current default auth store (root user, root role) to proxy and get it merged into master? So I can make a similar design based on yours later. And I think we can be parallel. When you finish the proxy part, I can make another PR for further integration.

heyitsanthony commented:
@hsluoyz 

> What you said about proxy may be a long-term goal, but I didn't see any efforts towards it though.

There's no effort yet because 1) it's not a high priority and 2) the KV API didn't support nested transactions until recently. That low priority isn't a mandate for accepting new code that will probably be a pain/timesink. Built-in auth is already difficult to mirror, its bugs can take down the entire cluster, and it's become forced overhead that tries to touch everything. Given that experience, I'm not confident this is a reliable way to build out auth support.

I'd prefer to avoid obvious design problems in the early stages if possible. Likewise, there's no reason to rush here as far as I can tell since there's already an adequate auth system in place.

> The current design mixed authentication and authorization and they should be separated sooner or later.

Can you open an issue about this? I'd like to know how this would fit in with the existing auth system and RPCs. Some discussion first would be helpful. Thanks!

heyitsanthony commented:
@hsluoyz I'm going to try hacking out the auth by proxy stuff tomorrow so you're not blocked; something experimental shouldn't take too long...

hsluoyz commented:
@heyitsanthony 

> Given that experience, I'm not confident this is a reliable way to build out auth support.

The Casbin auth store is not enabled by default. So you won't be affected if you don't want to use it. You will be still using the default auth. Of course, I will make sure Casbin auth store works without problems.

> there's no reason to rush here as far as I can tell since there's already an adequate auth system in place.

The same words. Please make a PR about your proposal, and get it merged into master. Make it work for the default auth. So I will rebase my work based on the master.

> Can you open an issue about this? I'd like to know how this would fit in with the existing auth system and RPCs. Some discussion first would be helpful. Thanks!

I don't have time to do it for now, as I need to finish this PR here first.

mitake commented:
Hi @hsluoyz 

> I think it has simplified the setup because we don't need to enforce to create a root role, and grant root role to root user.

The behaviour comes from v2's auth API and I think keeping it would be helpful for users who are already using the auth feature. Keeping the difference between the default auth store and casbin store small is important whichever the casbin feature is implemented in etcdserver or proxy.

hsluoyz commented:
Hi @mitake ,

> The behaviour comes from v2's auth API and I think keeping it would be helpful for users who are already using the auth feature. Keeping the difference between the default auth store and casbin store small is important whichever the casbin feature is implemented in etcdserver or proxy.

There's no difference between default auth store and casbin store for etcd users. Just the internal implementation is a little different. Let's see it:

- Default store

1. MUST create root user.
2. MUST create root role.
3. MUST grant root role to root user, so root user can do anything.
4. OPTIONAL root role can be granted to other users.

The final effect is root user and users with root role can do anything.

- Casbin store

1. MUST create root user, root user can do anything.
2. OPTIONAL create root role.
3. OPTIONAL root role can be granted to other users.

The final effect is also: root user and users with root role can do anything.

MUST is something we must do before enabling the auth. OPTIONAL is something the user can choose to do or do not.

So we can see the effects for both stores are the same. It exposes the same usage of API for etcd users. So they don't feel any difference in it.

The gain of Casbin store is that we don't need to setup too many things: only need to create a root user. The default store has to do three things: create root user, create root role, and grant root role to root user. This is tedious and also brings more checkings. For example, we have to forbid a user to revoke the root role from the root user. We have to forbid the user from deleting the root role. These checkings have to be done for every related operation, which may slow down the performance a bit.

hsluoyz commented:
Hi @mitake , is this PR still in progress? Or I need to wait for the job from @heyitsanthony ?

Hi @heyitsanthony ,

> I'm going to try hacking out the auth by proxy stuff tomorrow so you're not blocked; something experimental shouldn't take too long...

I will be grateful if you don't block this PR. But correct me if I'm wrong, I'm afriaid that's just what you are doing now.. Because I visited your link here: https://github.com/coreos/etcd/issues/8217 and your [forked repo](https://github.com/heyitsanthony/etcd/tree/auth-emu). it seems it is still WIP and hasn't been committed any code for 22 days. I really hope this PR can be in progress (or shutdown if the moderators feel not acceptable:)).

So what do you say? Thanks.

heyitsanthony commented:
@hsluoyz I'm not interested in merging this into the core etcd server, especially if the intent is to have it diverge from the current auth behavior while still appearing to serve the current auth APIs. Even if it's not a default, it'll still have to be supported and I'm not convinced it's worth the complexity. I suggested proxying because that would isolate casbin faults (and migration issues, etc) from the server, at least.

I haven't had time to finish the vauth stuff because, like I said, it's not a priority.

hsluoyz commented:
Thanks for the clarification!
